File: apps/plant_shop_c-sharp/plant_shop_c-sharp/Repositories/OrderRepository.cs
using Npgsql;
using plant_shop_c_sharp.Models;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace plant_shop_c_sharp.Repositories
{
    /// <summary>
    /// Repository CRUD pour les commandes.
    /// </summary>
    public class OrderRepository : BaseRepository
    {
        /// <summary>
        /// Constructeur avec injection de la source de donnees.
        /// </summary>
        /// <param name="dataSource">Source Npgsql.</param>
        public OrderRepository(NpgsqlDataSource dataSource) : base(dataSource) { }

        /// <summary>
        /// Trouve une commande par ID.
        /// </summary>
        /// <param name="id">ID a rechercher.</param>
        /// <returns>Commande ou null.</returns>
        public async Task<Order?> FindByIdAsync(int id)
        {
            await using var conn = GetConnection();
            await using var cmd = new NpgsqlCommand("SELECT * FROM orders WHERE id = @id", conn);
            cmd.Parameters.AddWithValue("id", id);

            await using var reader = await cmd.ExecuteReaderAsync();
            if (await reader.ReadAsync())
            {
                return MapOrder(reader);
            }
            return null;
        }

        /// <summary>
        /// Liste les commandes d un utilisateur.
        /// </summary>
        /// <param name="userId">ID de l utilisateur.</param>
        /// <returns>Liste des commandes.</returns>
        public async Task<List<Order>> FindByUserIdAsync(int userId)
        {
            var orders = new List<Order>();
            await using var conn = GetConnection();
            // Trié par date de création (plus récent en premier)
            await using var cmd = new NpgsqlCommand("SELECT * FROM orders WHERE user_id = @userId ORDER BY created_at DESC", conn);
            cmd.Parameters.AddWithValue("userId", userId);

            await using var reader = await cmd.ExecuteReaderAsync();
            while (await reader.ReadAsync())
            {
                orders.Add(MapOrder(reader));
            }
            return orders;
        }

        /// <summary>
        /// Cree une nouvelle commande.
        /// </summary>
        /// <param name="order">Commande a creer.</param>
        /// <returns>Commande avec ID genere.</returns>
        public async Task<Order> CreateAsync(Order order)
        {
            await using var conn = GetConnection();
            var sql = "INSERT INTO orders (user_id, total, status, created_at) VALUES (@userId, @total, @status, @created) RETURNING id";
            await using var cmd = new NpgsqlCommand(sql, conn);

            cmd.Parameters.AddWithValue("userId", order.UserId);
            cmd.Parameters.AddWithValue("total", order.TotalPrice);
            cmd.Parameters.AddWithValue("status", order.Status);
            cmd.Parameters.AddWithValue("created", DateTime.UtcNow);

            order.Id = (int)await cmd.ExecuteScalarAsync();
            return order;
        }

        /// <summary>
        /// Met a jour une commande.
        /// </summary>
        /// <param name="order">Commande modifiee.</param>
        public async Task UpdateAsync(Order order)
        {
            await using var conn = GetConnection();
            var sql = "UPDATE orders SET total = @total, status = @status WHERE id = @id";
            await using var cmd = new NpgsqlCommand(sql, conn);

            cmd.Parameters.AddWithValue("id", order.Id);
            cmd.Parameters.AddWithValue("total", order.TotalPrice);
            cmd.Parameters.AddWithValue("status", order.Status);

            await cmd.ExecuteNonQueryAsync();
        }

        /// <summary>
        /// Supprime une commande.
        /// </summary>
        /// <param name="id">ID a supprimer.</param>
        public async Task DeleteAsync(int id)
        {
            await using var conn = GetConnection();
            // S'assurer que les items sont supprimés (CASCADE ou manuellement)
            // Supposons CASCADE
            await using var cmd = new NpgsqlCommand("DELETE FROM orders WHERE id = @id", conn);
            cmd.Parameters.AddWithValue("id", id);
            await cmd.ExecuteNonQueryAsync();
        }

        /// <summary>
        /// Mappe un reader vers une Order.
        /// </summary>
        /// <param name="reader">Reader Npgsql.</param>
        /// <returns>Entite Order.</returns>
        private Order MapOrder(NpgsqlDataReader reader)
        {
            return new Order
            {
                Id = reader.GetInt32(reader.GetOrdinal("id")),
                UserId = reader.GetInt32(reader.GetOrdinal("user_id")),
                TotalPrice = reader.GetDecimal(reader.GetOrdinal("total")),
                Status = reader.GetString(reader.GetOrdinal("status")),
                CreatedAt = reader.GetDateTime(reader.GetOrdinal("created_at"))
            };
        }
    }
}

File: apps/plant_shop_java/Structure_en_application_distribuée/plant_shop_java_distribuée/order-service/src/repository/OrderRepository.java
import java.math.BigDecimal;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.List;

/**
 * Repository de base pour le service de commandes.
 * @param <T> Type de l'entité
 */
abstract class OrderBaseRepository<T> {
    protected final Connection db;
    protected final String table;

    /**
	 * Constructeur.
	 * @param db Connexion à la base de données
	 * @param table Nom de la table
	 */
	OrderBaseRepository(Connection db, String table) {
        this.db = db;
        this.table = table;
    }

    /**
	 * Mappe un ResultSet vers une entité.
	 * @param rs Le ResultSet
	 * @return L'entité mappée
	 */
	abstract T map(ResultSet rs) throws SQLException;

    /**
	 * Trouve une entité par son identifiant.
	 * @param id L'identifiant
	 * @return L'entité ou null
	 */
	T find(int id) throws SQLException {
        try (PreparedStatement ps = db.prepareStatement("SELECT * FROM " + table + " WHERE id=?")) {
            ps.setInt(1, id);
            try (ResultSet rs = ps.executeQuery()) {
                if (rs.next()) {
                    return map(rs);
                }
            }
        }
        return null;
    }

    /**
	 * Supprime une entité.
	 * @param id L'identifiant
	 */
	void delete(int id) throws SQLException {
        try (PreparedStatement ps = db.prepareStatement("DELETE FROM " + table + " WHERE id=?")) {
            ps.setInt(1, id);
            ps.executeUpdate();
        }
    }
}

/**
 * Repository pour les commandes.
 */
public final class OrderRepository extends OrderBaseRepository<Order> {

    /**
	 * Constructeur.
	 * @param db Connexion à la base de données
	 */
	public OrderRepository(Connection db) {
        super(db, "orders");
    }

    /**
	 * Mappe un ResultSet vers un objet Order.
	 * @param rs Le ResultSet
	 * @return La commande mappée
	 */
	@Override
    Order map(ResultSet rs) throws SQLException {
        return new Order(
            rs.getInt("id"),
            rs.getInt("user_id"),
            rs.getBigDecimal("total"),
            rs.getString("status"),
            rs.getTimestamp("created_at").toInstant()
        );
    }

    /**
	 * Crée une nouvelle commande.
	 * @param order La commande à créer
	 * @return L'identifiant généré
	 */
	public int create(Order order) throws SQLException {
        try (PreparedStatement ps = db.prepareStatement(
            "INSERT INTO orders(user_id, total, status) VALUES (?, ?, ?)",
            PreparedStatement.RETURN_GENERATED_KEYS)) {
            ps.setInt(1, order.userId);
            ps.setBigDecimal(2, order.total);
            ps.setString(3, order.status);
            ps.executeUpdate();
            try (ResultSet rs = ps.getGeneratedKeys()) {
                rs.next();
                return rs.getInt(1);
            }
        }
    }

    /**
	 * Met à jour le total d'une commande.
	 * @param id L'identifiant de la commande
	 * @param total Le nouveau total
	 */
	public void updateTotal(int id, BigDecimal total) throws SQLException {
        try (PreparedStatement ps = db.prepareStatement("UPDATE orders SET total=? WHERE id=?")) {
            ps.setBigDecimal(1, total);
            ps.setInt(2, id);
            ps.executeUpdate();
        }
    }

    /**
	 * Met à jour le statut d'une commande.
	 * @param id L'identifiant de la commande
	 * @param status Le nouveau statut
	 */
	public void updateStatus(int id, String status) throws SQLException {
        try (PreparedStatement ps = db.prepareStatement("UPDATE orders SET status=? WHERE id=?")) {
            ps.setString(1, status);
            ps.setInt(2, id);
            ps.executeUpdate();
        }
    }

    /**
	 * Liste les commandes d'un utilisateur.
	 * @param userId L'identifiant de l'utilisateur
	 * @return La liste des commandes
	 */
	public List<Order> listByUser(int userId) throws SQLException {
        List<Order> out = new ArrayList<>();
        try (PreparedStatement ps = db.prepareStatement("SELECT * FROM orders WHERE user_id=?")) {
            ps.setInt(1, userId);
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) {
                    out.add(map(rs));
                }
            }
        }
        return out;
    }
}
